FEATURE: Side-by-side comparison of changes in Workspace Management

robertlemke opened the pull request:
This change adds a side-by-side view of changes of a given workspace
to the "Review" view of the Workspace Management module. It highlights
inline text changes and displays thumbnails of changed assets.

This new feature relies on a third party diff library which has been
re-packaged, moved to a proper namespace, made Composer auto loading
compatible and been made compatible with PHP 7. It is from now on
part of the Neos distribution.

NEOS-956 #close

robertlemke commented:
![screen shot 2015-12-15 at 12 26 12](https://cloud.githubusercontent.com/assets/95582/11809425/190e7c1e-a327-11e5-8239-e0d03c2ae7bd.png)

I marked this changes as work-in-progress because there are a few details still missing:

- property labels are not translated yet
- some minor style glitches

However, please feel free to review this feature already!

kdambekalns commented on Neos.Diff/Classes/Diff.php:
```diff
@@ -0,0 +1,145 @@
+     * @param array $b Array containing the lines for the second string to compare.
+     * @param array $options Options (see $defaultOptions in this class)
+     */
+    public function __construct($a, $b, array $options = [])
```
Why do `$a` and `$b` not have a type hint?

robertlemke replied:
because the original library didn't have any. I only improved *some* of the code, and was rather happy that it would run on PHP 7. Well, that's how it is with external libraries.

But sure, let's improve it - now or later.

dimaip replied:
Aren't there any libraries on packagist that you could just require? Maintaining our own fork of some general diff algorithm doesn't seem cool.

kdambekalns replied:
I asked the same, but Robert spent some time looking yesterday, and there are libraries out there, yes. Some haven't been touched since *199x* - it's not a field of constant innovation… I guess the shortcut taken (adopt something) is the best option right now.

kdambekalns commented on Neos.Diff/Classes/Diff.php:
```diff
@@ -0,0 +1,145 @@
+     * Render a diff using the supplied rendering class and return it.
+     *
+     * @param Renderer\AbstractRenderer $renderer An instance of the rendering object to use for generating the diff.
+     * @return mixed The generated diff. Exact return value depends on the rendered.
```
depends on the rendered _what_?

robertlemke replied:
Not a clue.

kdambekalns commented on Neos.Diff/Classes/Renderer/Html/HtmlArrayRenderer.php:
```diff
@@ -0,0 +1,200 @@
+     * @param array $lines Array of lines to format.
+     * @return array Array of the formatted lines.
+     */
+    private function formatLines($lines)
```
array type hint?

robertlemke replied:
If you like ...

On one occasion I tried to add an array type hint elsewhere in the code and it turned out that it *sometimes* also was a string rather than an array. Well.

kdambekalns commented on Neos.Diff/Classes/SequenceMatcher.php:
```diff
@@ -0,0 +1,723 @@
+     * when calling the calculation methods, we need to recalculate them.
+     *
+     * @param string|array $b The sequence to set as the second sequence.
+     */
```
@return void

robertlemke replied:
@beautify code.

Oh and everyone reading this: Karsten also has a track record in reviewing minified JavaScript ;-) SCNR

kdambekalns commented on Neos.Diff/Classes/SequenceMatcher.php:
```diff
@@ -0,0 +1,723 @@
+     *
+     * @param string|array $a A string or array containing the lines to compare against.
+     * @param string|array $b A string or array containing the lines to compare.
+     */
```
@return void

kdambekalns commented on Neos.Diff/Classes/SequenceMatcher.php:
```diff
@@ -0,0 +1,723 @@
+        $this->setSequences($a, $b);
+    }
+
+    public function setOptions($options)
```
Docblock

kdambekalns commented on Neos.Diff/Classes/SequenceMatcher.php:
```diff
@@ -0,0 +1,723 @@
+     * Generate the internal arrays containing the list of junk and non-junk
+     * characters for the second ($b) sequence.
+     */
+    private function chainB()
```
@param and @return?

kdambekalns commented on Neos.Diff/Classes/SequenceMatcher.php:
```diff
@@ -0,0 +1,723 @@
+     * for the list of junk characters.
+     *
+     * @return boolean $b True if the character is considered junk. False if not.
+     */
```
@return void

kdambekalns commented on Neos.Diff/Classes/SequenceMatcher.php:
```diff
@@ -0,0 +1,723 @@
+     * @param array $triple Array containing the matching block triple to add to the running total.
+     * @return int The new running total for the number of matches.
+     */
+    private function ratioReduce($sum, $triple)
```
array type hint for `$triple`?

kdambekalns commented on Neos.Diff/Classes/SequenceMatcher.php:
```diff
@@ -0,0 +1,723 @@
+     * @param mixed $default The value to return as the default value if the key doesn't exist.
+     * @return mixed The value from the array if the key exists or otherwise the default.
+     */
+    private function arrayGetDefault($array, $key, $default)
```
array type hnt for `$array`

kdambekalns commented on Neos.Diff/Classes/SequenceMatcher.php:
```diff
@@ -0,0 +1,723 @@
+     * @param array $b Second array to compare.
+     * @return int -1, 0 or 1, as expected by the usort function.
+     */
+    private function tupleSort($a, $b)
```
array type hints

dimaip commented:
This feature is a must have, we should die but get it merged tonight!

Small nitpick: light gray on dark gray is damn hard to read. Currently we have two modes of text: white on #323232 or black on white when the text field is active in selector.
I think the most appropriate for this situation is to use black font on white background, to set apart the content from the interface, and make it a lot more readable.

kdambekalns commented:
One comment to begin with: styling and missing labels should not block this from being merged, IMHO.

Then, sorry for the nitpicks on the diff library, but… keep it clean from the start.

Now on to reviewing the actual change to Neos…

robertlemke commented:
@dimaip I tried it with a white background and it looked pretty ugly in the context of the workspace management module. The grey on dark grey is on purpose of course, since its only there to provide context and should stand back behind the actual text changes. But you're certainly right that it could be a bit lighter.

kdambekalns commented on TYPO3.Neos/Classes/TYPO3/Neos/Controller/Module/Management/WorkspacesController.php:
```diff
@@ -491,6 +506,144 @@ protected function computeSiteChanges(Workspace $selectedWorkspace)
+        if (!isset($this->contextContexts[$baseWorkspaceName])) {
+            $contextProperties = $modifiedNode->getContext()->getProperties();
+            $contextProperties['workspaceName'] = $baseWorkspaceName;
+            $this->contextContexts[$baseWorkspaceName] = $this->contextFactory->create($contextProperties);
```
If my workspace contains changes to nodes having differing dimension values, the first node will populate the cache and the second would have it's original node fetched from the "wrong" workspace… no?

robertlemke replied:
Fixed in latest commit

kdambekalns commented on TYPO3.Neos/Classes/TYPO3/Neos/Controller/Module/Management/WorkspacesController.php:
```diff
@@ -491,6 +506,144 @@ protected function computeSiteChanges(Workspace $selectedWorkspace)
+    }
+
+    /**
+     *
```
Either drop the empty lines, or (better) write an obvious one-line description…

kdambekalns commented on TYPO3.Neos/Classes/TYPO3/Neos/Controller/Module/Management/WorkspacesController.php:
```diff
@@ -491,6 +506,144 @@ protected function computeSiteChanges(Workspace $selectedWorkspace)
+                $changedSlimmedDownContent = $this->renderSlimmedDownContent($changedPropertyValue);
+
+                $diff = new Diff(explode("\n", $originalSlimmedDownContent), explode("\n", $changedSlimmedDownContent),
+                    ['context' => 1]);
```
Hm, either all parameters on their own line, or all in one line.

robertlemke replied:
yes, PhpStorm reformatted that, need to check my style rules.

dimaip commented:
@robertlemke yeah I understand your intent, that the unchanged text is less important than the changed text, but it should still be legible. The contrast should be increased A LOT for the text to become readable, especially on cheap office LCD screens, not everyone has a retina nowadays...

kdambekalns commented:
Ok, added some more nitpicks and a potential real problem. Otherwise looks good, and I can take care of fixing the nitpicks, if needed. An actual test is to follow.

kdambekalns commented:
Note: I don't (usually) review minified JS…

kdambekalns commented on Neos.Diff/Classes/SequenceMatcher.php:
```diff
@@ -0,0 +1,723 @@
+        }
+
+        while ($bestI > $alo && $bestJ > $blo && $this->isBJunk($b[$bestJ - 1]) &&
+            !$this->isLineDifferent($bestI - 1, $bestJ - 1)) {
```
should probably also be `linesAreDifferent`, because `isLineDifferent` doesn't exist…

robertlemke commented:
FYI, an updated version is in the works ... with improved contrast ;-)

kdambekalns commented:
Did some testing and it works just fine. And I had a sneek peek on the new look, totally works for me. :)

Let's get this in after robert pushed the latest version and do some more polishing (labels, styling) until the release. :tada:

robertlemke commented:
![image](https://cloud.githubusercontent.com/assets/95582/11817271/81352202-a354-11e5-980e-79f7d0080fc8.png)

Coming up ...

dimaip commented:
@robertlemke your last screenshot just made my day.
Ping me when the latest stuff is ready for testing.
Currently playing with the current version, seems to work great.

robertlemke commented:
Okay, pushed the version you can see in the screenshot. What's now missing is the translatable property labels. I could imagine that we add that tomorrow, even after the feature freeze, but then again, could be tonight, too.

kdambekalns commented:
Latest version works for me, good job!

@robertlemke my comments on the `WorkspacesController` are still valid

robertlemke commented:
![screen shot 2015-12-15 at 18 59 13](https://cloud.githubusercontent.com/assets/95582/11819037/0e845a20-a35e-11e5-95d6-9066d911db6b.png)
![screen shot 2015-12-15 at 18 59 36](https://cloud.githubusercontent.com/assets/95582/11819038/0e856ca8-a35e-11e5-8711-50cd7c9d8bc2.png)

Screenshots of the latest version. This looks good (enough) for me now, only contains some minor style glitches which we can iron out later.

robertlemke commented:
@dimaip ping ;-)

aertmann commented:
looks great! agree there are some details that needs to be improved but can be done afterwards..

however I have some questions..

I take it the colored border replaces the legend (new, modified, deleted, etc.)

the node label is not displayed anymore, only the node type which is secondary information when identifying an element.. I'd suggest to re-add the label instead of the node type label and display the node type label more discrete or using a tooltip..

additionally I don't see the relative path anymore, is that removed or hidden somewhere? again here it could be shown using a tooltip to save space..

last I fear that lots of changes or long texts or lots of changed properties won't work very well.. maybe the diff should be hidden by default and opened manually?
curious how it would work in practice when doing lots of changes before publishing

kitsunet commented:
Very nice :)

IMHO something we can fix after merging but I am not quite sure why the second example doesn't work:

Change is very small but ok, it's highlighted, probably nothing we can do to make that more visible...
![screen shot 2015-12-15 at 21 00 59](https://cloud.githubusercontent.com/assets/324408/11822511/2394a670-a36f-11e5-87d0-3d52f6b901e5.png)

But here if you compare left to right you will see one removal and one addition but both are not highlighted:
![screen shot 2015-12-15 at 21 01 07](https://cloud.githubusercontent.com/assets/324408/11822510/239213b0-a36f-11e5-87d3-c373dd3810c0.png)

kitsunet commented on TYPO3.Neos/Resources/Private/Translations/en/Modules.xlf:
```diff
@@ -137,8 +137,8 @@
 			</trans-unit>
-			<trans-unit id="workspaces.legendCaption" xml:space="preserve">
-				<source>Legend</source>
+			<trans-unit id="changedContentCaption" xml:space="preserve">
```
This would be one of the few changes that have no prefix (workspaces) does that make sense? Shouldn't it have one?

robertlemke replied:
sure, that makes sense and slipped through, will fix it in a minute.

kitsunet commented on TYPO3.Neos/Classes/TYPO3/Neos/Controller/Module/Management/WorkspacesController.php:
```diff
@@ -491,6 +500,146 @@ protected function computeSiteChanges(Workspace $selectedWorkspace)
+                        'diff' => $diffArray
+                    ];
+                }
+            } elseif ($originalPropertyValue instanceof AssetInterface || $changedPropertyValue instanceof AssetInterface) {
```
Fine for now but this needs more flexibility for other objects. IMHO we shouldn't have a direct Asset reference in here.

robertlemke replied:
agreed; in general we really need to extract the diff functionality into a service once we start implementing more features.

robertlemke commented:
@aertmann thanks for your feedback!

> I take it the colored border replaces the legend (new, modified, deleted, etc.)

yes, I also added a little tooltip / title field which is visible when you hover over the change. That could be improved by adding more than just the "delete" or "moved" term which I re-used from the previous legend.

> the node label is not displayed anymore, only the node type which is secondary information when identifying an element.. I'd suggest to re-add the label instead of the node type label and display the node type label more discrete or using a tooltip..

I removed the node label because in the majority of all the content I tried it with, it was generated from the first chunks of content, so you would end up with labels like `Don't explain anymore how editing works. Just discover it yourself! All contents are directly edited on the page, just styled l…`

We can add it again, but at least from the content I played with I didn't see a big benefit. On the other hand, only using the node type icon was too confusing, because many custom node types in customer projects supposedly use similar icons (also because we are limited to the ones existing in Font Awesome).

> additionally I don't see the relative path anymore, is that removed or hidden somewhere? again here it could be shown using a tooltip to save space..

The relative path of the content elements were not really telling much as the node names a completely random (for me they only looked nice or added some information when I used the demo site, because we created the nodes at a time when the identifiers were not random).

What I could imagine as an improvement for later is grouping content by first-level content collections.

> last I fear that lots of changes or long texts or lots of changed properties won't work very well.. maybe the diff should be hidden by default and opened manually? curious how it would work in practice when doing lots of changes before publishing

I tried it with starting with collapsed changes, but it's a hassle to go through the content changes when you review then. I think we need to get more experience with real-world scenarios (how large will review sets will be?) and then could add further navigation (a customized tree) or, for example, some search function. Also lazy-loading will be needed if you are working with large amounts of content.

kitsunet commented on TYPO3.Neos/Classes/TYPO3/Neos/Controller/Module/Management/WorkspacesController.php:
```diff
@@ -491,6 +500,146 @@ protected function computeSiteChanges(Workspace $selectedWorkspace)
+     * Renders a slimmed down representation of a property of the given node. The output will be HTML, but does not
+     * contain any markup from the original content.
+     *
+     * Note: It's clear that this method needs to be extracted and moved to a more universal service at some point.
```
Will be interesting for the History module I guess.

kitsunet commented:
:+1: I like so far.

robertlemke commented:
There's one further issue with the positioning of the Discard / Publish button in the fixed footer:

![image](https://cloud.githubusercontent.com/assets/95582/11822929/7f6d3d66-a371-11e5-982d-1376aaf378b9.png)

and when I styled the review view, I broke the table in the workspace overview page:

![image](https://cloud.githubusercontent.com/assets/95582/11822963/bef6effe-a371-11e5-8a42-2a2cc4bc7671.png)

(see mispositioned `Review` buttons)

robertlemke commented:
LOL, the `Review` button is not misplaced because of a styling issue, but because a round-off-error which results in less dots ...

kitsunet commented:
hihi, but IMHO that big of an issue. Could still be fixed int he next days.

kitsunet commented:
Did you see my comment above with the screenshots?

robertlemke commented:
@kitsunet yes, it's pretty weird why the changes are not detected. Since the text per se is displayed, it looks like a glitch in the diff library. I'll try to reproduce and debug it here.

kdambekalns commented:
:+1: From looking at it, I'm fine with merging this. The remaining issues can be ironed out the next days, and further refactorings be done at a later stage for one of the next versions.

I'll update my local copy and test again, and will merge unless I find something or there is an objection here within the next few minutes. :)

robertlemke commented:
@kdambekalns thanks! But I do have a little improvement here which is almost done, just wait a couple of minutes, I'll ping you.

kdambekalns commented:
Ok, no problem…

robertlemke commented:
I improved the change stats in the workspace overview: they don't have rounding errors anymore, and more importantly, the bars are not so confusing anymore when you have few changes (with the "blocks" based stats, you'd see 5 blocks in orange even if you only modified one content element).

![image](https://cloud.githubusercontent.com/assets/95582/11824146/479ebade-a378-11e5-8ce0-0804cd171fb1.png)

robertlemke commented:
Okay, good to go - I couldn't track reproduce @kitsunet's issue right now, but it shouldn't be a big deal to fix that later this week.

:+1:

kdambekalns commented:
:+1: - change stats look good now!

dimaip commented:
I'm a bit late for the party, but here's a small note for future improvements: I think the review button gets a bit lost here. Would make sense to make the whole row clickable, and only have the trash button do the trashing.
![diff](https://cloud.githubusercontent.com/assets/837032/11826958/ba6299b6-a399-11e5-9384-79f68df53b71.png)

dimaip commented:
One more thing: currently the shortest way to get to this nice review mode from normal editing mode happens in three click: publish menu dropdown->workspaces->review button.
Maybe we could add a shortcut item to publish menu with something like: "review my unpublished changes" that would jump straight to review mode for my own changes?

dimaip commented:
@robertlemke @aertmann what do you think about my last comment? I tried this feature on our editors, and they all voiced the same concerns, that they would want a quick shortcut path to overview of their own changes before publishing.
You know, it's just like you do in git, always review your own changes before pushing.

kitsunet commented:
Indeed would make a lot of sense to me if that was possible from the publish menu directly ("review current workspace").

And yes I somewhat agree that the "Review" could get a button styling.

aertmann commented:
@dimaip fully agree, did some stuff with the review button to make it more visible.. also it's now only shown if there are any changes

and :+1: for adding a direct link in the publish menu.. thought about it yesterday, but couldn't think of a good description for it so left it out

will implement it now with the suggestion from @kitsunet

robertlemke commented:
I think that it's important that the buttons are styled consistently. So if its not clear that it is a button through the hover effect alone, we should also check other places where we have buttons and re-style them accordingly. I think, what's viable is to show an icon next to the non-button, like @aertmann suggested in his other change.

As for the button being visible at all if there are no changes: I have a rather strong opinion about this and would like that the button is not visible at all if there are no changes in the workspace. You would not look for that button if there weren't any changes, and as soon as there are changes, you would understand that there's a button for reviewing them. It's a different case for the trash icon: even if you're not allowed to delete a certain workspace, there should be a deactivated trash button which gives you that hint: theoretically you could delete this, but the popover tells you why you can't do that at the moment.

robertlemke commented:
Ah, about the button / menu item in the Publish selector button: yes, let's try that. My suggestion for the label would be "Review changes"
